src/python/xml_parser-py.c

```diff
@@ -792,3 +790,247 @@ py_xml_parse_main_metadata_together(G_GNUC_UNUSED PyObject *self, PyObject *args
 }
+
+typedef struct {
+    PyObject_HEAD
+    cr_PkgIterator *pkg_iterator;
+} _PkgIteratorObject;
+
+cr_PkgIterator *
+PkgIterator_FromPyObject(PyObject *o)
+{
+    if (!PkgIteratorObject_Check(o)) {
+        PyErr_SetString(PyExc_TypeError, "Expected a createrepo_c.PkgIterator object.");
+        return NULL;
+    }
+    return ((_PkgIteratorObject *)o)->pkg_iterator;
+}
+
+PyObject *
+Object_FromPkgIterator(cr_PkgIterator *pkg_iterator)
+{
+    if (!pkg_iterator) {
+        PyErr_SetString(PyExc_ValueError, "Expected a cr_PkgIterator pointer not NULL.");
+        return NULL;
+    }
+
+    PyObject *py_pkg_iterator = PyObject_CallObject((PyObject *)&PkgIterator_Type, NULL);
+    ((_PkgIteratorObject *)py_pkg_iterator)->pkg_iterator = pkg_iterator;
+    return py_pkg_iterator;
+}
+
+static int
+check_PkgIteratorStatus(const _PkgIteratorObject *self)
+{
+    assert(self != NULL);
+    assert(PkgIteratorObject_Check(self));
+    if (self->pkg_iterator == NULL) {
+        PyErr_SetString(CrErr_Exception, "Improper createrepo_c PkgIterator object.");
+        return -1;
+    }
+    return 0;
+}
+
+/* Function on the type */
+
+static PyObject *
+pkg_iterator_new(PyTypeObject *type,
+            G_GNUC_UNUSED PyObject *args,
+            G_GNUC_UNUSED PyObject *kwds)
+{
+    _PkgIteratorObject *self = (_PkgIteratorObject *)type->tp_alloc(type, 0);
+    if (self) {
+        self->pkg_iterator = NULL;
+    }
+    return (PyObject *)self;
+}
+
+PyDoc_STRVAR(pkg_iterator_init__doc__,
+             "PkgIterator object\n\n"
+             ".. method:: __init__()\n\n"
+             "    Default constructor\n");
+
+static int
+pkg_iterator_init(_PkgIteratorObject *self, PyObject *args, PyObject *kwargs)
+{
+    char *primary_path;
+    char *filelists_path;
+    char *other_path;
+    PyObject *py_newpkgcb, *py_warningcb;
+    CbData cbdata;
+    GError *tmp_err = NULL;
+    static char *kwlist[] = {"primary", "filelists", "other", "newpkgcb",
+                             "warningcb", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "sssOO:pkg_iterator_init", kwlist,
+                                     &primary_path, &filelists_path, &other_path, &py_newpkgcb,
+                                     &py_warningcb)) {
+        return -1;
+    }
+
+    if (!PyCallable_Check(py_newpkgcb) && py_newpkgcb != Py_None) {
+        PyErr_SetString(PyExc_TypeError, "newpkgcb must be callable or None");
+        return -1;
+    }
+
+    if (!PyCallable_Check(py_warningcb) && py_warningcb != Py_None) {
+        PyErr_SetString(PyExc_TypeError, "warningcb must be callable or None");
+        return -1;
+    }
+
+    if (self->pkg_iterator) // reinitialization by __init__()
+        cr_PkgIterator_free(self->pkg_iterator, &tmp_err);
+    // TODO: correct thing to do with error?
+    if (tmp_err) {
+        nice_exception(&tmp_err, NULL);
+        return -1;
+    }
+
+    Py_XINCREF(py_newpkgcb);
+    Py_XINCREF(py_warningcb);
+    Py_XINCREF(cbdata.py_pkgs);
```

Comment:
This line definitely shouldn't be here. The `cbdata.py_pkgs` dictionary is not even created yet and I don't think we need to do this anyway.
```suggestion
```